ykpack/src/types.rs

```diff
@@ -439,6 +449,83 @@ pub enum Statement {
     Unimplemented(String),
 }
 
+impl Statement {
+    /// Returns a vector of locals that this SIR statement *may* define.
+    /// Whether or not the local is actually defined depends upon if this is the first write into
+    /// the local (there is no explicit liveness marker in SIR/TIR).
+    pub fn maybe_defined_locals(&self) -> Vec<Local> {
+        let mut ret = Vec::new();
+
+        match self {
+            Statement::Nop => (),
+            Statement::Assign(place, _rval) => place.push_maybe_defined_locals(&mut ret),
+            // `Enter` doesn't define the destination, as that will be defined by an inlined assignment.
+            Statement::Enter(_target, args, _dest_place, start_idx) => {
+                for idx in 0..args.len() {
+                    // + 1 to skip return value.
+                    ret.push(Local(start_idx + u32::try_from(idx).unwrap() + 1));
+                }
+            }
+            Statement::Leave => (),
+            Statement::StorageDead(_) => (),
+            Statement::Call(_target, _args, dest) => {
+                if let Some(dest) = dest {
+                    dest.push_maybe_defined_locals(&mut ret);
+                }
+            }
+            Statement::Unimplemented(_) => (),
+        }
+
+        ret
+    }
+
+    /// Returns a vector of locals that this SIR statement uses but does not define.
+    pub fn used_locals(&self) -> Vec<Local> {
+        let mut ret = Vec::new();
+
+        match self {
+            Statement::Nop => (),
+            Statement::Assign(place, rval) => {
+                rval.push_used_locals(&mut ret);
+                place.push_used_locals(&mut ret);
+            }
+            // `Enter` doesn't use the callee args. Inlined statements will use them instead.
+            Statement::Enter(_target, _args, _opt_place, _idx) => (),
+            Statement::Leave => (),
+            Statement::StorageDead(_) => (),
+            Statement::Call(_target, args, _dest) => {
+                for a in args {
+                    a.push_used_locals(&mut ret);
+                }
+            }
+            Statement::Unimplemented(_) => (),
+        }
+
+        ret
+    }
+
+    /// Returns a vector of locals that this SIR statement removes the existing definitions of.
+    /// The locals are considered dead at the start of the following statement.
```

Comment:
s/following/next? Otherwise one could interpret this as "the statement following this comment".